RP2040 fails to build with newer framework versions

### The problem

Hi,

the default framework version is quite old and contains bugs I need to avoid.
I was able to change to framework version some time ago, for example:

```
rp2040:
  board: rpipico
  framework:
    platform_version: https://github.com/maxgerhardt/platform-raspberrypi.git
    version: 4.5.2
```

However this stopped working as an item was renamed:

```
Compiling .pioenvs\yambms3\src\esphome\components\rp2040\gpio.cpp.o
src\esphome\components\rp2040\gpio.cpp: In member function 'void esphome::ISRInternalGPIOPin::pin_mode(esphome::gpio::Flags)':
src\esphome\components\rp2040\gpio.cpp:111:22: error: 'padsbank0_hw' was not declared in this scope; did you mean 'pads_bank0_hw'?
  111 |     hw_write_masked(&padsbank0_hw->io[arg->pin],
      |                      ^~~~~~~~~~~~
      |                      pads_bank0_hw
*** [.pioenvs\yambms3\src\esphome\components\rp2040\gpio.cpp.o] Error 1
```

Now I can just change that from `padsbank0_hw` to `pads_bank0_hw` and it compiles/works again fine. Even with latest 4.7.0 framework.

```
INFO ESPHome 2025.7.5
INFO Reading configuration yambms3_Shunts.yaml...
WARNING The selected Arduino framework version is not the recommended one.
WARNING The selected Arduino framework version is not the recommended one.
INFO Generating C++ source...
INFO Compiling app...
Processing yambms3 (board: rpipico; framework: arduino; platform: https://github.com/maxgerhardt/platform-raspberrypi.git)
------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Verbose mode can be enabled via `-v, --verbose` option
CONFIGURATION: https://docs.platformio.org/page/boards/raspberrypi/rpipico.html
PLATFORM: Raspberry Pi RP2040 (1.16.0+sha.1f67f1e) > Pico
HARDWARE: RP2040 133MHz, 256KB RAM, 2MB Flash
DEBUG: Current (blackmagic) External (blackmagic, cmsis-dap, jlink, picoprobe, raspberrypi-swd)
PACKAGES:
 - framework-arduinopico @ 1.40700.0 (4.7.0)
 - tool-picotool-rp2040-earlephilhower @ 5.140200.240929 (14.2.0)
 - toolchain-rp2040-earlephilhower @ 5.140200.240929 (14.2.0)
Flash size: 2.00MB
Sketch size: 1.00MB
Filesystem size: 1.00MB
PSRAM size: 0.00MB
Maximium Sketch size: 1044480 EEPROM start: 0x101ff000 Filesystem start: 0x100ff000 Filesystem end: 0x101ff000
LDF: Library Dependency Finder -> https://bit.ly/configure-pio-ldf
LDF Modes: Finder ~ chain+, Compatibility ~ strict
Found 69 compatible libraries
Scanning dependencies...
No dependencies
Building in release mode
Compiling .pioenvs\yambms3\src\esphome\components\rp2040\gpio.cpp.o
Retrieving maximum program size .pioenvs\yambms3\firmware.elf
Flash size: 2.00MB
Sketch size: 1.00MB
Filesystem size: 1.00MB
PSRAM size: 0.00MB
Maximium Sketch size: 1044480 EEPROM start: 0x101ff000 Filesystem start: 0x100ff000 Filesystem end: 0x101ff000
Checking size .pioenvs\yambms3\firmware.elf
Advanced Memory Usage is available via "PlatformIO Home > Project Inspect"
RAM:   [          ]   4.5% (used 11820 bytes from 262144 bytes)
Flash: [=         ]  14.0% (used 146348 bytes from 1044480 bytes)
Building .pioenvs\yambms3\firmware.bin.signed
=========================================================================================================================== [SUCCESS] Took 4.48 seconds ===========================================================================================================================
INFO Successfully compiled program.
```

I could provide a PR for that change, but I don't know who will change the recommended framework version? Any tips on where to start?

### Which version of ESPHome has the issue?

2025.7.5

### What type of installation are you using?

pip

### What platform are you using?

RP2040

### Component causing the issue

gpio

### YAML Config

```yaml

```

### Anything in the logs that might be useful for us?

```txt

```

### Additional information

_No response_

--- a/esphome/components/adc/adc_sensor_rp2040.cpp
+++ b/esphome/components/adc/adc_sensor_rp2040.cpp
@@ -8,6 +8,13 @@
 #endif  // CYW43_USES_VSYS_PIN
 #include <hardware/adc.h>
 
+// PICO_VSYS_PIN is defined in pico-sdk board headers (e.g. boards/pico2.h),
+// but the Arduino framework's config_autogen.h includes a generic board header
+// that doesn't define it. Provide the standard value (pin 29) as a fallback.
+#ifndef PICO_VSYS_PIN
+#define PICO_VSYS_PIN 29  // NOLINT(cppcoreguidelines-macro-usage)
+#endif
+
 namespace esphome {
 namespace adc {
 

--- a/tests/components/adc/test.rp2040-pico2-ard.yaml
+++ b/tests/components/adc/test.rp2040-pico2-ard.yaml
@@ -1,0 +1,11 @@
+sensor:
+  - id: my_sensor
+    platform: adc
+    pin: VCC
+    name: ADC Test sensor
+    update_interval: "1:01"
+    unit_of_measurement: "°C"
+    icon: "mdi:water-percent"
+    accuracy_decimals: 5
+    setup_priority: -100
+    force_update: true

--- a/tests/components/spi/test.rp2040-pico2-ard.yaml
+++ b/tests/components/spi/test.rp2040-pico2-ard.yaml
@@ -1,0 +1,6 @@
+substitutions:
+  clk_pin: GPIO2
+  mosi_pin: GPIO3
+  miso_pin: GPIO4
+
+<<: !include common.yaml

--- a/tests/test_build_components/build_components_base.rp2040-pico2-ard.yaml
+++ b/tests/test_build_components/build_components_base.rp2040-pico2-ard.yaml
@@ -1,0 +1,15 @@
+esphome:
+  name: componenttestrp2040pico2ard
+  friendly_name: $component_name
+
+rp2040:
+  board: rpipico2
+
+logger:
+  level: VERY_VERBOSE
+
+packages:
+  component_under_test: !include
+    file: $component_test_file
+    vars:
+      component_test_file: $component_test_file

--- a/tests/test_build_components/common/spi/rp2040-pico2-ard.yaml
+++ b/tests/test_build_components/common/spi/rp2040-pico2-ard.yaml
@@ -1,0 +1,12 @@
+# Common SPI configuration for RP2040 Pico 2 (RP2350) Arduino tests
+
+substitutions:
+  clk_pin: GPIO18
+  mosi_pin: GPIO19
+  miso_pin: GPIO16
+
+spi:
+  - id: spi_bus
+    clk_pin: ${clk_pin}
+    mosi_pin: ${mosi_pin}
+    miso_pin: ${miso_pin}

